tests/test.rs

```diff
@@ -151,6 +153,219 @@ async fn test_transaction() {
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    use std::sync::{Arc, Mutex};
+
+    let transaction_attempted_count = Arc::new(Mutex::new(0));
+    let should_retry_query_count = Arc::new(Mutex::new(0));
+
+    // Test a transaction that:
+    //
+    // 1. Retries on the first call
+    // 2. Explicitly rolls back on the second call
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+    let err = conn
+        .transaction_async_with_retry(
+            |_conn| {
+                let transaction_attempted_count = transaction_attempted_count.clone();
+                async move {
+                    let mut count = transaction_attempted_count.lock().unwrap();
+                    *count += 1;
+
+                    if *count < 2 {
+                        eprintln!("test: Manually restarting txn");
+                        return Err::<(), _>(diesel::result::Error::DatabaseError(
+                            diesel::result::DatabaseErrorKind::SerializationFailure,
+                            Box::new("restart transaction".to_string()),
+                        ));
+                    }
+                    eprintln!("test: Manually rolling back txn");
+                    return Err(diesel::result::Error::RollbackTransaction);
+                }
+            },
+            || async {
+                *should_retry_query_count.lock().unwrap() += 1;
+                true
+            },
+        )
+        .await
+        .expect_err("Transaction should have failed");
+
+    assert_eq!(err, diesel::result::Error::RollbackTransaction);
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+
+    // The transaction closure should have been attempted twice, but
+    // we should have only asked whether or not to retry once -- after
+    // the first failure, but not the second.
+    assert_eq!(*transaction_attempted_count.lock().unwrap(), 2);
+    assert_eq!(*should_retry_query_count.lock().unwrap(), 1);
+
+    test_end(crdb).await;
+}
+
+#[tokio::test]
+async fn test_transaction_automatic_retry_injected_errors() {
+    let crdb = test_start().await;
+
+    let manager = async_bb8_diesel::ConnectionManager::<PgConnection>::new(&crdb.pg_config().url);
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    use std::sync::{Arc, Mutex};
+
+    let transaction_attempted_count = Arc::new(Mutex::new(0));
+    let should_retry_query_count = Arc::new(Mutex::new(0));
+
+    // Tests a transaction that is forced to retry by CockroachDB.
+    //
+    // By setting this session variable, we expect that:
+    // - "any statement executed inside of an explicit transaction (with the
+    // exception of SET statements) will return a transaction retry error."
+    // - "after the 3rd retry error, the transaction will proceed as
+    // normal"
```

Comment:
A link to [here](https://www.cockroachlabs.com/docs/v23.1/transaction-retry-error-example#test-transaction-retry-logic) would probably be helpful. I didn't understand where the magic of the 3rd attempt comes from, but it's part of setting that session variable in CRDB.

Reply:
Link added!